File: examples/form-interaction.mjs
/**
 * Webcompat repro pattern: form or button interaction fails in Firefox
 *
 * Replace the URL, selectors, and expected outcome with the site under investigation.
 * Run: ff-bidi-run examples/form-interaction.mjs --firefox /path/to/firefox --headless
 */
import { startSession } from 'ff-test-firefox-bidi-client';

const TARGET_URL = 'https://example.com';
const INPUT_SELECTOR = 'input[type="text"]';
const SUBMIT_SELECTOR = 'button[type="submit"]';
const SUCCESS_SELECTOR = '.result';

const session = await startSession();
const { page } = session;

try {
  await page.goto(TARGET_URL);

  await page.evaluate(([sel, val]) => {
    const el = document.querySelector(sel);
    el.focus();
    el.value = val;
    el.dispatchEvent(new Event('input', { bubbles: true }));
    el.dispatchEvent(new Event('change', { bubbles: true }));
  }, [INPUT_SELECTOR, 'test value']);
  await page.click(SUBMIT_SELECTOR);

  // Wait up to 5 seconds for the result to appear
  try {
    await page.waitForSelector(SUCCESS_SELECTOR, { timeout: 5000 });
    console.log('issue not reproduced');
  } catch {
    const url = await page.url();
    console.log(`issue reproduced: expected result element not found after interaction (url: ${url})`);

    // Save a screenshot for debugging
    await page.screenshot({ path: 'repro-screenshot.png' });
    console.log('  screenshot saved to repro-screenshot.png');
  }
} finally {
  await session.close();
}

File: python/examples/form_interaction.py
"""
Webcompat repro pattern: form or button interaction fails in Firefox

Use this when you observed: filling a form and clicking submit produced no result,
an error, or a wrong outcome — while the same steps worked in another browser.

Adjust the URL, selectors, input value, and success condition to what you observed.
"""

import asyncio
import json
from firefox_bidi_client import start_session

TARGET_URL = "https://example.com"
INPUT_SELECTOR = 'input[type="text"]'
SUBMIT_SELECTOR = 'button[type="submit"]'
SUCCESS_SELECTOR = ".result"


async def main():
    async with await start_session() as session:
        page = session.page
        await page.goto(TARGET_URL)

        # Fill the input field
        await page.evaluate(
            "([sel, val]) => {"
            "  const el = document.querySelector(sel);"
            "  el.focus();"
            "  el.value = val;"
            "  el.dispatchEvent(new Event('input', { bubbles: true }));"
            "  el.dispatchEvent(new Event('change', { bubbles: true }));"
            "}",
            [INPUT_SELECTOR, "test value"],
        )

        await page.click(SUBMIT_SELECTOR)

        try:
            await page.wait_for_selector(SUCCESS_SELECTOR, timeout=5.0)
            print("issue not reproduced")
        except TimeoutError:
            url = await page.url()
            print(f"issue reproduced: expected result element not found after interaction (url: {url})")
            await page.screenshot("repro-screenshot.png")
            print("  screenshot saved to repro-screenshot.png")


asyncio.run(main())
